binding.c

```diff
@@ -92,40 +98,93 @@ parse_address (struct sockaddr *name, char *ip, size_t size, int *port, int *fam
   }
 }
 
+/**
+ * When a native function is executing on uv event;
+ * (cannot be surrounded by try-catch block)
+ *
+ * redirect pending exceptions to global uncaught handler
+ * to avoid process abort (given an uncaught handler is installed).
+ */
+static inline void
+redirect_to_uncaught (js_env_t *env, int err) {
+  if (err == 0) return;
+  assert(err == js_pending_exception || err == js_uncaught_exception);
+
+  bool is_pending;
+  err = js_is_exception_pending(env, &is_pending);
+  assert(err == 0);
+
+  if (is_pending) {
+    js_value_t *fatal_exception;
+    err = js_get_and_clear_last_exception(env, &fatal_exception);
+    assert(err == 0);
+
+    err = js_fatal_exception(env, fatal_exception);
+    assert(err == 0);
+  }
+}
```

Comment:
This should never be necessary as it's handled by the compatibility layer: https://github.com/holepunchto/bare-compat-napi/blob/a50f09ae6b7f35f834e375c398dbde076baf2bd4/include/js.h#L1322-L1340. If it's currently necessary then we're likely missing a microtask checkpoint somewhere.

Reply:
You are entirely correct. I ported a previous sanity check to it's own repo:
[test.js](https://github.com/telamon/native-addon/blob/0c15e3cfbc7fe588986d56826005fda9d469b67b/test.js#L32)
[binding.c](https://github.com/telamon/native-addon/blob/0c15e3cfbc7fe588986d56826005fda9d469b67b/binding.c#L95)

And compared to old notes i must have accidentally tested exception handling on an i/o call without creating a checkpoint. *sigh*

Key takeaway:
- Exceptions are not allowed to escape to outer calls
- There are no pending exceptions on outer calls. (don't ever call `js_get_and_clear_last_exception()` in this context)

Removing function above